onmt/translate/BeamSearcher.lua

```diff
@@ -0,0 +1,395 @@
+      :contiguous():view(batchSize, 1)
+      :expand(batchSize, beamSize):contiguous():view(-1))
+  end)
+end
+
+--[[Helper function. Recursively convert flat `batchSize * beamSize` tensors
+ to 2D `(batchSize, beamSize)` tensors.
+
+Parameters:
+
+  * `v` - flat tensor of size `batchSize * beamSize` or a table containing such tensors.
+  * `beamSize` - beam size
+
+Returns: `(batchSize, beamSize)` tensor or a table containing such tensors.
+
+--]]
+local function flatToRc(v, beamSize)
+  return onmt.utils.Tensor.recursiveApply(v, function (h)
+    local batchSize = math.floor(h:size(1) / beamSize)
+    local sizes = {}
+    for j = 2, #h:size() do
+        sizes[j - 1] = h:size(j)
+    end
+    return h:view(batchSize, beamSize, table.unpack(sizes))
+  end)
+end
+
+--[[Helper function. Recursively convert 2D `(batchSize, beamSize)` tensors to
+ flat `batchSize * beamSize` tensors.
+
+Parameters:
+
+  * `v` - flat tensor of size `(batchSize, beamSize)` or a table containing such tensors.
+  * `beamSize` - beam size
+
+Returns: `batchSize * beamSize` tensor or a table containing such tensors.
+
+--]]
+local function rcToFlat(v)
+  return onmt.utils.Tensor.recursiveApply(v, function (h)
+    local sizes = {}
+    sizes[1] = h:size(1) * h:size(2)
+    for j = 3, #h:size() do
+        sizes[j - 1] = h:size(j)
+    end
+    return h:view(table.unpack(sizes))
+  end)
+end
+
+--[[Helper function. Recursively select `(batchSize, ...)` tensors by
+  specified batch indexes.
+
+Parameters:
+
+  * `v` - tensor of size `(batchSize, ...)` or a table containing such tensors
+  * `indexes` - a table of the desired batch indexes
+
+Returns: Indexed `(newBatchSize, ...)` tensor or a table containing such tensors
+
+--]]
+local function selectBatch(v, remaining)
+  return onmt.utils.Tensor.recursiveApply(v, function (h)
+    if not torch.isTensor(remaining) then
+      remaining = torch.LongTensor(remaining)
+    end
+    return h:index(1, remaining)
+  end)
+end
+
+--[[Helper function. Recursively select `(batchSize * beamSize, ...)` tensors by
+  specified batch index and beam index.
+
+Parameters:
+
+  * `v` - tensor of size `(batchSize * beamSize, ...)` or a table containing such tensors
+  * `beamSize` - beam size
+  * `batch` - the desired batch index
+  * `beam` - the desired beam index
+
+Returns: Indexed `(...)` tensor or a table containing such tensors
+
+--]]
+local function selectBatchBeam(v, beamSize, batch, beam)
+  return onmt.utils.Tensor.recursiveApply(v, function (h)
+    local batchSize = math.floor(h:size(1) / beamSize)
+    local sizes = {}
+    for j = 2, #h:size() do
+        sizes[j - 1] = h:size(j)
+    end
+    local hOut = h:view(batchSize, beamSize, table.unpack(sizes))
+    return hOut[{batch, beam}]
+  end)
+end
+
+
+--[[Helper function.
+
+Parameters:
+
+  * `v` - tensor
+  * `x` - reference tensor
+
+Returns: if `x` is cuda tensor, return `v:cuda()`; otherwise, return `v`.
+
+--]]
+local function localize(v, x)
```

Comment:
Previously I was using `onmt.utils.Cuda.convert`, but as pointed out by your previous comment, that function might not know if we are in gpu mode or not, since it needs to be initialized by `opt`. (The reason I had to convert back and forth is that CudaByteTensor does not support some functions that CudaDoubleTensor supports.)